logstash-core/lib/logstash/util/wrapped_acked_queue.rb

```diff
@@ -191,23 +184,9 @@ def add_output_metrics(filtered_size)
 
     class ReadBatch
       def initialize(queue, size, wait)
-        @queue = queue
-        @size = size
-        @wait = wait
-
-        @originals = Hash.new
-
-        # TODO: disabled for https://github.com/elastic/logstash/issues/6055 - will have to properly refactor
-        # @cancelled = Hash.new
-
         @generated = Hash.new
-        @acked_batch = nil
-      end
-
-      def read_next
-        @acked_batch = @queue.read_batch(@size, @wait)
-        return if @acked_batch.nil?
-        @acked_batch.get_elements.each { |e| @originals[e] = true }
+        @acked_batch = queue.read_batch(size, wait)
+        @originals = @acked_batch.nil? ? Hash.new : @acked_batch.get_elements
```

Comment:
`@originals` and `@generated` are both `Hash{Object=>TrueClass}`, which are only used for inclusion detection and iteration; functionally they're using a custom `Set` implementation;

Instead of changing `RubyAckedBatch#get_elements` to leak this particular detail, should we instead make it return a `Set` (either a Java `Set` or a `RubySet`) and refactor this class' internals to use the appropriate methods for inclusion detection and iteration?

Reply:
@yaauie there isn't really a fast `RubySet` and a Java Set (also note that we need an ordered set here) will be very slow here until we do the next steps (because of JRuby wrapping it in a Java Proxy).
=> for now, this is the best we can do here I fear, but "now" won't last long I promise :) Hopefully, we'll get the next steps done here (including moving to Java collections) this week :)